### DISCUSSION TITLE
Filter combining inconsistency between blockers

### DISCUSSION THREAD

**Discussion by peace2000** [uBlockOrigin/uBlock-issues#1728] [Interoperability]
I have noticed that there is an inconsistency between blockers when combining filters with a comma (and there's a procedural operator involved).

Adblock Plus, Adguard and uBlock Origin treat filters like this differently:

`iltalehti.fi#?#main div[class="card "], main .half-article:-abp-contains(Kaupallinen yhteistyö)`

ABP & uBO way to interpret:

`iltalehti.fi#?#main div[class="card "]:-abp-contains(Kaupallinen yhteistyö)`
`iltalehti.fi#?#main .half-article:-abp-contains(Kaupallinen yhteistyö)`

AG way to interpret:
`iltalehti.fi#?#main div[class="card "]`
`iltalehti.fi#?#main .half-article:-abp-contains(Kaupallinen yhteistyö)`

I'm not aware of any common place to discuss stuff like this so I'm opening an issue report here, and I'll tag representatives from each "section": @gorhill @ameshkov @WebReflection

What I'm trying to say is that could you consider doing some alignment/agreeing together how rules like that should be interpret so that it's interpret in the same way in all blockers? For those who try to maintain cross-blocker compatible filterlists, inconsistencies in interpretation are not a good thing.

Page where I tested filter interpretation: `https://www.iltalehti.fi/raha`

**Comment by ghost** [comment]
https://github.com/uBlockOrigin/uBlock-issues/issues/382#issuecomment-703735587

***

Actually this one issue was linked from wiki: https://github.com/uBlockOrigin/uBlock-issues/issues/453 it's about comma-separated procedurals, not only `:style()`.

**Comment by gorhill** [comment]
> I'm not aware of any common place to discuss stuff like this 

Please open a [discussion](https://github.com/uBlockOrigin/uBlock-issues/discussions) about this, I want the issue tracker to be kept for actionable items.

**Comment by ghost** [comment]
Thread moved to discussions.

**Comment by ameshkov** [comment]
AG follows CSS selectors semantics here. I don't fully understand why uBO & ABP opted to do it differently.

On a side note, does it really make sense to support commas in CSS selectors? Combining different selectors may save some time, but it hurts our attempts to automate the detection of obsolete selectors.

**Reply by gorhill** [reply]
> I don't fully understand why uBO & ABP opted to do it differently.

I just didn't want to write a full CSS parser. AdGuard is using Sizzle, so this comes for free, but at the expense of having to inject a sizable amount of content script code.

**Reply by ameshkov** [reply]
In the future, we'd like to replace Sizzle with a parser of our own and publish it as a separate library. Indeed, we don't need half of the functionality Sizzle provides. Would you consider using it as a dependency if it matches your needs?

**Comment by uBlock-user** [comment]
> does it really make sense to support commas in CSS selectors?

Very useful where multiple elements are to be styled with one common style or get removed or something similar where all elements are subjected to one singular action.

> Adblock Plus, Adguard and uBlock Origin treat filters like this differently:

No, procedural filters cannot be combined with plain selectors like that in uBO.

**Reply by peace2000** [reply]
> > Adblock Plus, Adguard and uBlock Origin treat filters like this differently:
> 
> No, procedural filters cannot be combined with plain selectors like that in uBO.

Yes, the point of that rule is to combine two procedurals.

**Comment by peace2000** [comment]
The rule I brought forward was made by a certain uBO user (not @uBlock-user :D) who suggested me to combine two procedural filters that way.

I didn't add that rule to the Finnish Easylist because I luckily happened to be aware of this syntax interpretation difference. Also, I don't usually like to combine rules (with the exception of style injection or if rules are targeting visually the same element). But someone else filterlist maintainer who uses uBO or ABP, could add a rule like that and it could cause breakages to AG users.

The rule I gave in the opening, would hide legit articles on `iltalehti.fi` in AG but not in uBO or ABP.

**Reply by ameshkov** [reply]
It just bothers me that in order to be compatible here we would need to change the semantically correct logic to an incorrect one. We'll be keeping an eye on this. Of course, if the issues like that continue to occur, we'll have to somehow react, but I'd prefer not to change AG logic here.

**Comment by DandelionSprout** [comment]
I recall being involved in a discussion about this a couple months ago, though I can no longer remember where it occured. I find the uBO+ABP approach to be ***very*** convenient, as some of my opt-in cultural filterlists would not have been realistically possible to maintain without `example.com##.element,.element2,.element3:has-text(Text)`.

The filesize of [this specific list](https://raw.githubusercontent.com/DandelionSprout/adfilt/master/Anti-F%D1%96%D0%9C%20List.txt) would've possibly doubled with AdGuard's approach, though I'm short on time to do fine calculations about it this afternoon.

**Reply by ameshkov** [reply]
Regardless of how convenient it is, it's semantically incorrect. I find it hard to explain to a new filter maintainer why `:not` and `:has` work differently.

Consider one more thing. At some point, some of the procedural selectors may become a part of the CSS standard. For instance, this will soon happen with `:has`, it will soon be supported by Chrome natively. What will happen to all those rules then?

**Reply by DandelionSprout** [reply]
I'll gladly admit that I'm not an expert on the official CSS foundation standards, with me at least *traditionally* also being confused when the ABP coding team argued that they want to stick as close to the official standards as possible.

So my concerns lay solely with filesize, and with how many lines I'd need to add/edit when adding entries against new keywords.

**Reply by krystian3w** [reply]
Possible use `:is()` and for outdated browsers:

`:-moz-any()` / `:-webkit-any()` / `:matches()`

To try combine only one `not`.

IMO uBo no implemented translate modern `:is()` construction to legacy CSS name in outdated browser  (user like update only addons and by "noscripts" tribe try disable known 0-day bugs).

**Comment by Yuki2718** [comment]
@krystian3w It's better to rewrite Sep'21 idea of FB Sponsored in `https://www.reddit.com/r/uBlockOrigin/wiki/solutions`

**Reply by krystian3w** [reply]
I am shadowbanned so ask gwarser / "RraaLL" / ublock user to remove outdated filters.

**Reply by krystian3w** [reply]
Also I not copied these into "Polskie filtry elementów irytujących" so only outdated "My Filters" tab may break Facebook.

**Reply by u-RraaLL** [reply]
> It's better to rewrite Sep'21 idea of FB Sponsored in `https://www.reddit.com/r/uBlockOrigin/wiki/solutions`

Did you mean so it'd match AG-syntax?
While I don't think it's necessary on account of uBO's reddit wiki not being a filter list, I coincidentally fixed that issue when I updated the solution last week.

**Reply by Yuki2718** [reply]
@u-RraaLL Not necessary, just a bit better. Nice to see you here.

**Reply by gorhill** [reply]
> I am shadowbanned

I don't know what you are talking about, there is no shadowbanning on GitHub.

**Reply by u-RraaLL** [reply]
@gorhill  He meant on reddit. His account got deleted/banned in January (IIRC).

@Yuki2718 Thanks.

**Comment by krystian3w** [comment]
This may changed a few logic with now pure CSS filters: https://github.com/uBlockOrigin/uBlock-issues/issues/2480 so better check how works own lists where is not added long time ago `:is(..., ...):has(...)` or splited to many lines with duplicate `:has(...)`.

Demo to checks: `jsbin.com,net##form, td:has(b)`

https://jsfiddle.net/wvcnz8Lk/ https://output.jsbin.com/tedikod
